Add a function to simply refresh the grains

terminalmage opened the pull request:
Also use this function to refresh the grains when editing them using
grains.{delval,setval,setvals}, avoiding a full sync. This speeds up
these functions by skipping the syncing of custom grains modules. For
most use cases, walking all your fileserver environments (which could be
a lot if using gitfs) is unnecessary when you're simply using these
functions to add/remove grains from /etc/salt/grains.

terminalmage committed:
Add a function to simply refresh the grains

Also use this function to refresh the grains when editing them using
grains.{delval,setval,setvals}, avoiding a full sync. This speeds up
these functions by skipping the syncing of custom grains modules. For
most use cases, walking all your fileserver environments (which could be
a lot if using gitfs) is unnecessary when you're simply using these
functions to add/remove grains from /etc/salt/grains.

terminalmage commented:
@cro Since part of this code inovles whether or not the minion is a proxy, I'd like you to look at the changes I made in minion.py and make sure they won't negatively impact proxy minions.

cro commented on salt/minion.py:
```diff
@@ -1685,6 +1685,15 @@ def _fire_master_minion_start(self):
+        self.opts['grains'] = salt.loader.grains(
+            self.opts,
+            force_refresh=True,
+            proxy=salt.utils.is_proxy())
```
`proxy` here is expecting a salt.loader.proxy reference but salt.utils.is_proxy() returns True or False.

terminalmage replied:
Ahh. should I just be using ``self.proxy`` here then?

terminalmage replied:
or rather, ``getattr(self, 'proxy', None)`` ?

cro replied:
Yes, that'll do it.

terminalmage replied:
Done. Thank you sir!

terminalmage committed:
Properly hand proxy minions

The proxy argument to salt.loader.grains expects a reference to a
salt.loader.proxy loader instance, not a bool. Thanks @cro!

terminalmage commented:
I made the changes recommended by @cro, this should be good to merge when tests complete.

terminalmage commented:
This still has some changes that need to be made in the mocking for unit tests which were mocking the results of ``saltutil.sync_grains`` and now must be updated to instead mock for ``saltutil.refresh_grains``. I'll take care of this over the weekend.

terminalmage committed:
Fix mocking for grains refresh

terminalmage commented:
Go Go Jenkins!

terminalmage commented:
Go Go Jenkins!

terminalmage commented:
Should be good to merge now.
